Refactor to keep ref vs. alt allele reads

iskandr opened the pull request:
Fixes: 
* https://github.com/hammerlab/isovar/issues/30
* https://github.com/hammerlab/isovar/issues/25

`VariantRead` has been replaced with `AlleleRead` which can be optionally filtered into the subset of reads which support a variant allele. 

Some of these changes are also preparing for weighted RNA expression scores using partially matching reads (https://github.com/hammerlab/isovar/issues/26) and integration with Topiary.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/hammerlab/isovar/31)
<!-- Reviewable:end -->

tavinathanson commented:
From @iskandr:

> Changes to isovar: 
* got rid of `VariantRead`, now have `AlleleRead` which can be optionally filtered to particular alleles (like ref vs. alt)
* Refactored the args into several sub-modules to mirror what I’m doing with Topiary, hoping to eliminate some redundant arg parsing stuff between Topiary, Isovar, and Vaxrank. 
* Changed some functions which starts from VCF + BAM to instead take a generator of variants + supporting reads, cuts down on duplicated args and makes it easier to preserve overlapping/alt/cdna-sequence-supporting/protein-sequence-supporting reads in different spots
* Lots of renames, using “overlapping” to describe reads of any allele and “supporting” to describe reads with the same allele as a variant.

tavinathanson commented on isovar/allele_reads.py:
```diff
@@ -13,89 +13,52 @@
 
     Parameters
     ----------
     read : ReadAtLocus
```
Rename `read` to `read_at_locus` in the docstring?

iskandr replied:
👍

tavinathanson commented on test/test_protein_sequences.py:
```diff
@@ -96,10 +96,12 @@ def make_dummy_protein_sequence(
-        total_variant_reads=n_total_variant_reads,
-        supporting_transcripts=[None] * n_supporting_reference_transcripts,
-        total_transcripts=n_total_reference_transcripts,
+        overlapping_reads=[None] * n_supporting_variant_reads,
```
`n_total_variant_reads` no longer used; remove from function?

iskandr replied:
That's actually a mistake, the number of alt reads should be n_total_variant_reads

tavinathanson commented on script/isovar-allele-reads.py:
```diff
@@ -24,29 +24,28 @@
+    df = allele_reads_dataframe(
+        variants,
+        samfile,
+        min_mapping_quality=args.min_mapping_quality)
```
Was `min_mapping_quality` just forgotten before? Was that a bug?

tavinathanson commented on isovar/variant_sequences.py:
```diff
@@ -0,0 +1,254 @@
+# Copyright (c) 2016. Mount Sinai School of Medicine
```
Pure rename or any differences of note?

iskandr replied:
There are actually a lot of small changes, I didn't realize the file rename would mask those.

tavinathanson commented on isovar/variant_reads.py:
```diff
@@ -0,0 +1,53 @@
+    _, _, alt = trim_variant(variant)
+    return [read for read in allele_reads if read.allele == alt]
+
+def filter_non_alt_reads_for_variants(variants_and_allele_reads_sequence):
```
Where is this used?

iskandr replied:
Immediately below in `reads_supporting_variants` and anywhere where you want to have both `allele_reads` and `variant_reads` simultaneously.

tavinathanson replied:
I specifically mean `filter_non_alt_reads_for_variants`, not `filter_non_alt_reads_for_variant`

iskandr replied:
Huh, I guess it used to be used in `protein_sequences` but now I'm doing it via groupby_allele instead. I think it's still worth keeping in case someone only cares about total vs. alt reads.

tavinathanson commented on isovar/string_helpers.py:
```diff
@@ -0,0 +1,55 @@
+import logging
+
+
+def trim_N_nucleotides(prefix, suffix):
```
Any changes here, or just a move?

iskandr replied:
Just a move.

tavinathanson commented:
Reviewed 5 of 48 files at r1, 1 of 5 files at r2, 6 of 27 files at r4.
Review status: 12 of 51 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[.travis.yml, line 47 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbCTqV3luleQN91FqQ:-KLbCTqV3luleQN91FqR:-259697923) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/.travis.yml#L47)):*
> ```YAML
>   - ./lint.sh
>   # run tests
>   - nosetests test --with-coverage --cover-package=isovar
> ```

Really wish we could apply this to all projects at once!

---

*[isovar/allele_reads.py, line 49 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbHs_dFu8P82T-OkSY:-KLbHs_dFu8P82T-OkSZ:910863149) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/allele_reads.py#L49)):*
> ```Python
>     ----------
>     locus_read : LocusRead
>         Read which may possibly contain the alternate nucleotides
> ```

Comment shouldn't be alternate specific now

---

*[isovar/allele_reads.py, line 51 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbHgI-OYUUbz3y2Koq:-KLbHgI0FjjMagV9Ibt6:-1659213299) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/allele_reads.py#L51)):*
> ```Python
>         Read which may possibly contain the alternate nucleotides
> 
>     n_ref : int
> ```

I'm confused about why this was added vs. using `len(ref)`, is it just that `ref` is unnecessary?

---

*[isovar/allele_reads.py, line 117 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbLaoQu0UbuvkI1vV2:-KLbLaoQu0UbuvkI1vV3:1780917270) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/allele_reads.py#L117)):*
> ```Python
> 
>     nucleotides_at_variant_locus = sequence[read_pos_before + 1:read_pos_after]
> ```

Note to self: this is where `nucleotides_at_variant_locus != alt` is dropped

---

*[isovar/variant_reads.py, line 24 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbNsgmtoo6rVhJZdea:-KLbNsgmtoo6rVhJZdeb:1801125009) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/variant_reads.py#L24)):*
> ```Python
> def filter_non_alt_reads_for_variant(variant, allele_reads):
>     _, _, alt = trim_variant(variant)
>     return [read for read in allele_reads if read.allele == alt]
> ```

Note to self: this is where the `alt` filtering moved to

---

*[isovar/args/protein_sequences.py, line 31 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbVC55XQJUk4zA2ZwE:-KLbVC55XQJUk4zA2ZwF:-1451815461) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/protein_sequences.py#L31)):*
> ```Python
>     Extends an ArgumentParser instance with the following args:
>         --max-protein-sequences-per-variant
>     Also adds all translation arguments such as:
> ```

Maybe add: see `args.tranlation`?

---

*[isovar/args/protein_sequences.py, line 51 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbVXhZOlWbvwy20a7t:-KLbVXhZOlWbvwy20a7u:1924484938) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/protein_sequences.py#L51)):*
> ```Python
>     return reads_generator_to_protein_sequences_generator(
>         allele_reads_generator,
>         protein_sequence_length=args.protein_sequence_length,
> ```

I'm nervous that somewhere in all this argparsing is a function parameter called with the wrong `args.argument`. Not sure how to feel less anxious about that, maybe testing the scripts? 

---

*[isovar/args/protein_sequences.py, line 54 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbVQrwU-h1a76aGRTF:-KLbVQrwU-h1a76aGRTG:862837552) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/protein_sequences.py#L54)):*
> ```Python
>         min_reads_supporting_cdna_sequence=args.min_reads_supporting_variant_sequence,
>         min_transcript_prefix_length=args.min_transcript_prefix_length,
>         max_transcript_mismatches=args.max_reference_transcript_mismatches,
> ```

Nit: a bit weird that some of these arg names don't match up with the params. Absolutely feel free to ignore.

---

*[isovar/args/reference_context.py, line 50 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbPYfv0VpV2kwNFDaq:-KLbPYfv0VpV2kwNFDar:580916723) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/reference_context.py#L50)):*
> ```Python
>             "a reference transcript. Values greater than zero exclude variants "
>             "near the start codon of transcrPROTEIN_SEQUENCE_LEGNTHipts without 5' UTRs."))
>     if add_context_size_arg:
> ```

Not clear to me why this arg is optionally vs. always added to the parser? (And if there is a good reason, should probably be commeted.)

---

*[isovar/args/reference_context.py, line 59 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbUQxXvb0iFKE3_beZ:-KLbUQxXvb0iFKE3_be_:149988926) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/reference_context.py#L59)):*
> ```Python
> def make_reference_context_arg_parser(add_context_size_arg=True, **kwargs):
>     parser = ArgumentParser(**kwargs)
>     add_somatic_vcf_args(parser)
> ```

Agreed (per your Slack comments): it's weird that you need to just remember this vs. something more automatic. No brilliant ideas on my side though.

---

*[isovar/args/rna_reads.py, line 59 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbWtgZvyr5KA-4Q5NH:-KLbWtgZvyr5KA-4Q5NI:222390916) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/rna_reads.py#L59)):*
> ```Python
>         "--drop-secondary-alignments",
>         default=True,
>         action="store_false")
> ```

Why `store_false` here? The `not args.drop_secondary_alignments` below looks like it wouldn't work right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hammerlab/isovar/31)*
<!-- Sent from Reviewable.io -->

tavinathanson commented:
Reviewed 7 of 48 files at r1, 3 of 27 files at r4, 1 of 7 files at r5.
Review status: 18 of 51 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[isovar/assembly.py, line 86 \[r5\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLd8t8rIt7q0NIliWr0:-KLd8t8s0WulIR2wOIqg:817972782) ([raw file](https://github.com/hammerlab/isovar/blob/7f9d421091a0cb78e4983e881cb05b949852cf9d/isovar/assembly.py#L86)):*
> ```Python
> def collapse_substrings(assembly_groups):
>     """
>     Combine shorter sequences which are fully contained in longer sequences
> ```

After 10 minutes trying to understand how this function works, I still don't fully. Can you comment it up more?

---

*[isovar/assembly.py, line 96 \[r5\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLd8zg3V6U8h8yLRMbn:-KLd8zg3V6U8h8yLRMbo:387426505) ([raw file](https://github.com/hammerlab/isovar/blob/7f9d421091a0cb78e4983e881cb05b949852cf9d/isovar/assembly.py#L96)):*
> ```Python
>         found_superstring = False
>         for (prefix_long, allele_long, suffix_long), names_long in result_list:
>             if allele != allele_long:
> ```

Can you explain this (and in a comment)?

---

*[isovar/dataframe_builder.py, line 117 \[r5\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLdA9fA18edoDyLrHQr:-KLdA9fBR_WAPDrHBZjJ:-1201334960) ([raw file](https://github.com/hammerlab/isovar/blob/7f9d421091a0cb78e4983e881cb05b949852cf9d/isovar/dataframe_builder.py#L117)):*
> ```Python
>                 value = fn(value)
> 
>             if isinstance(value, (tuple, list, set)):
> ```

Not clear to me what this is doing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hammerlab/isovar/31)*
<!-- Sent from Reviewable.io -->

iskandr commented:
Review status: 18 of 51 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[isovar/allele_reads.py, line 48 \[r1\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLS4_Khz0vj0dduYoss-r1-48:-KLcB8Jahr5vu6lvPOgH:-499146009) ([raw file](https://github.com/hammerlab/isovar/blob/0285a3aceaeca205dcc16c09b8cb4597259954d0/isovar/allele_reads.py#L48)):*
<details><summary><i>Previously, iskandr (Alex Rubinsteyn) wrote…</i></summary>
> 👍 
</details>
Done.

---

*[isovar/allele_reads.py, line 51 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbHgI-OYUUbz3y2Koq:-KLcBOaDHy05gQbO_ADt:1589247502) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/allele_reads.py#L51)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> I'm confused about why this was added vs. using `len(ref)`, is it just that `ref` is unnecessary?
</details>
Yeah, I'm trying to pass around minimal information.

---

*[isovar/dataframe_builder.py, line 117 \[r5\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLdA9fA18edoDyLrHQr:-KLhG8S8XwpiTGXEW4k8:-494958427) ([raw file](https://github.com/hammerlab/isovar/blob/7f9d421091a0cb78e4983e881cb05b949852cf9d/isovar/dataframe_builder.py#L117)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> Not clear to me what this is doing?
</details>
Almost all my uses of `converters` was to take the length of a collection, so I just made it the default behavior. 

---

*[isovar/args/protein_sequences.py, line 31 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbVC55XQJUk4zA2ZwE:-KLcD7jJuElTLFH2YfLp:-354349019) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/protein_sequences.py#L31)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> Maybe add: see `args.tranlation`?
</details>
Added docstrings to these parser constructors with references to their "parents"

---

*[isovar/args/protein_sequences.py, line 51 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbVXhZOlWbvwy20a7t:-KLcDGjxvWXe6gi5TpbZ:621945533) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/protein_sequences.py#L51)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> I'm nervous that somewhere in all this argparsing is a function parameter called with the wrong `args.argument`. Not sure how to feel less anxious about that, maybe testing the scripts? 
</details>
I actually changed some of the protein_sequences tests to use the commandline args instead of calling functions directly. 

Added an issue to test this stuff more thoroughly: https://github.com/hammerlab/isovar/issues/32

---

*[isovar/args/protein_sequences.py, line 54 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbVQrwU-h1a76aGRTF:-KLcDjeOB-mVoRq6jYdD:-1221890277) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/protein_sequences.py#L54)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> Nit: a bit weird that some of these arg names don't match up with the params. Absolutely feel free to ignore.
</details>
Yeah, I agree. There are actually a lot of small inconsistencies, added an issue: https://github.com/hammerlab/isovar/issues/33

---

*[isovar/args/reference_context.py, line 50 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbPYfv0VpV2kwNFDaq:-KLcDurFLArLKbPZesg3:-1654338776) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/reference_context.py#L50)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> Not clear to me why this arg is optionally vs. always added to the parser? (And if there is a good reason, should probably be commeted.)
</details>
Added a doc string sort of explaining but actually some of the commandline args should set this to False. Since this PR is already too big, I'm going to punt on that too:  https://github.com/hammerlab/isovar/issues/34

---

*[isovar/args/reference_context.py, line 59 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbUQxXvb0iFKE3_beZ:-KLcEKxU5kHqzRhfs_KG:-1755350165) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/reference_context.py#L59)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> Agreed (per your Slack comments): it's weird that you need to just remember this vs. something more automatic. No brilliant ideas on my side though.
</details>
I'm making a weird inheritance tree of parser constructors, each of which comes with its own deconstructor. This is begging for an abstraction of some sort, I just don't know exactly what. 

---

*[isovar/args/rna_reads.py, line 59 \[r4\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLbWtgZvyr5KA-4Q5NH:-KLcEXZIgvj4pn0_Rbv4:1698712105) ([raw file](https://github.com/hammerlab/isovar/blob/066d4ff93d0a889942782dd935347c714704e289/isovar/args/rna_reads.py#L59)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> Why `store_false` here? The `not args.drop_secondary_alignments` below looks like it wouldn't work right?
</details>
You're right, I got the values backwards. It should be `default=False, action="store_true"`. 

---

*[script/isovar-allele-reads.py, line 49 \[r1\]](https://reviewable.io:443/reviews/hammerlab/isovar/31#-KLS4_Ki-MsKopSFiqan-r1-49:-KLcEreMSpzrJctLTatT:677250776) ([raw file](https://github.com/hammerlab/isovar/blob/0285a3aceaeca205dcc16c09b8cb4597259954d0/script/isovar-allele-reads.py#L49)):*
<details><summary><i>Previously, tavinathanson (Tavi Nathanson) wrote…</i></summary>
> Was `min_mapping_quality` just forgotten before? Was that a bug?
</details>
Yeah

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/hammerlab/isovar/31)*
<!-- Sent from Reviewable.io -->
